File: 02-mapreduce/wordcount/src/main/java/ok/ml/WordCount.java
package ok.ml;

import java.io.IOException;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.conf.Configured;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.IntWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Partitioner;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;
import org.apache.hadoop.util.Tool;
import org.apache.hadoop.util.ToolRunner;

public class WordCount extends Configured implements Tool {
    public static final String IN_PATH_PARAM = "wordcount.input";
    public static final String OUT_PATH_PARAM = "wordcount.output";

    @Override
    public int run(String[] args) throws IOException, InterruptedException, ClassNotFoundException {
        Job job = Job.getInstance(getConf(), "Word Count");
        job.setJarByClass(WordCount.class);
        job.setMapperClass(Tokenizer.class);
        job.setReducerClass(IntSummator.class);
        job.setOutputKeyClass(Text.class);
        job.setOutputValueClass(IntWritable.class);
        job.setNumReduceTasks(2);
        job.setPartitionerClass(AlphabetPartitioner.class);
        FileInputFormat.addInputPath(job, new Path(getConf().get(IN_PATH_PARAM)));
        FileOutputFormat.setOutputPath(job, new Path(getConf().get(OUT_PATH_PARAM)));
        int exitCode = job.waitForCompletion(true) ? 0 : 1;
        System.out.println(job.getCounters().toString());
        return exitCode;
    }

    public static void main(String[] args) throws Exception {
        System.exit(ToolRunner.run(new Configuration(), new WordCount(), args));
    }

    public static class Tokenizer extends Mapper<Object, Text, Text, IntWritable> {
        private final IntWritable one = new IntWritable(1);
        private final Text text = new Text();

        @Override
        public void map(Object key, Text value, Context context) throws IOException, InterruptedException {
            String line = value.toString().toLowerCase();
            String[] words = line.split("\\s");
            for (String word : words) {
                text.set(word);
                context.write(text, one);
            }
        }
    }

    public static class IntSummator extends Reducer<Text, IntWritable, Text, IntWritable> {
        private final IntWritable result = new IntWritable();

        @Override
        public void reduce(Text key, Iterable<IntWritable> values, Context context) throws IOException, InterruptedException {
            int sum = 0;
            for (IntWritable val : values) {
                sum += val.get();
            }
            result.set(sum);
            context.write(key, result);
        }
    }

    public static class AlphabetPartitioner extends Partitioner<Text, IntWritable> {
        @Override
        public int getPartition(Text key, IntWritable value, int numPartitions) {
            if (numPartitions == 1) {
                return 0; // all in one
            }
            return (key.getLength() > 2) ? 1 : 0;
        }
    }
}

File: 02-mapreduce/wordcount-2/src/main/scala/WordCount.scala
package ok.ml

import org.apache.hadoop.conf.{Configuration, Configured}
import org.apache.hadoop.fs.Path
import org.apache.hadoop.io.{IntWritable, Text}
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat
import org.apache.hadoop.mapreduce.lib.reduce.IntSumReducer
import org.apache.hadoop.mapreduce.{Job, Mapper, Partitioner}
import org.apache.hadoop.util.{Tool, ToolRunner}

class Tokenizer extends Mapper[Object, Text, Text, IntWritable] {
  val one = new IntWritable(1)
  val text = new Text()

  override def map(key: Object, value: Text, context: Mapper[Object, Text, Text, IntWritable]#Context): Unit = {
    for (word <- value.toString.trim.toLowerCase.split("\\s")) {
      text.set(word)
      context.write(text, one)
    }
  }
}

class WordLenPartitioner extends Partitioner[Text, IntWritable] {
  override def getPartition(key: Text, value: IntWritable, numPartitions: Int): Int = numPartitions match {
    case 1 => 0
    case _ => if (key.getLength > 2) 1 else 0
  }
}

class WordCount extends Configured with Tool {
  val IN_PATH_PARAM = "wordcount.input"
  val OUT_PATH_PARAM = "wordcount.output"

  override def run(args: Array[String]): Int = {
    val job = Job.getInstance(getConf, "Word Count")
    job.setJarByClass(getClass)
    job.setOutputKeyClass(classOf[Text])
    job.setOutputValueClass(classOf[IntWritable])
    job.setMapperClass(classOf[Tokenizer])
    // use existing reducer
    job.setReducerClass(classOf[IntSumReducer[Text]])
    job.setNumReduceTasks(2)
    job.setPartitionerClass(classOf[WordLenPartitioner])
    FileInputFormat.addInputPath(job, new Path(getConf.get(IN_PATH_PARAM)))
    FileOutputFormat.setOutputPath(job, new Path(getConf.get(OUT_PATH_PARAM)))
    if (job.waitForCompletion(true)) 0 else 1
  }
}

object WordCount {
  def main(args: Array[String]): Unit =
    System.exit(ToolRunner.run(new Configuration, new WordCount, args))
}